File: bto_management_system/model/enquiry.py
from utils.input_util import InputUtil
from common.exceptions import DataLoadError, OperationError

class Enquiry:
    """Represents an enquiry submitted by an applicant."""
    _HEADERS = ['EnquiryID', 'ApplicantNRIC', 'ProjectName', 'Text', 'Reply']

    def __init__(self, enquiry_id: int, applicant_nric: str, project_name: str, text: str, reply: str = ""):
        if not isinstance(enquiry_id, int): raise ValueError("Enquiry ID must be an integer.")
        if not InputUtil.validate_nric(applicant_nric): raise ValueError("Invalid Applicant NRIC")
        if not project_name: raise ValueError("Project Name cannot be empty")
        if not text: raise ValueError("Enquiry text cannot be empty")

        self._enquiry_id = enquiry_id
        self._applicant_nric = applicant_nric
        self._project_name = project_name
        self._text = text
        self._reply = reply if reply is not None else ""

    # --- Getters ---
    @property
    def enquiry_id(self): return self._enquiry_id
    @property
    def applicant_nric(self): return self._applicant_nric
    @property
    def project_name(self): return self._project_name
    @property
    def text(self): return self._text
    @property
    def reply(self): return self._reply

    # --- State Checks ---
    def is_replied(self) -> bool:
        return bool(self._reply)

    # --- State Modifiers ---
    def set_text(self, new_text: str):
        if self.is_replied():
            raise OperationError("Cannot edit an enquiry that has already been replied to.")
        if not new_text:
            raise ValueError("Enquiry text cannot be empty.")
        self._text = new_text

    def set_reply(self, reply_text: str):
        if not reply_text:
             raise ValueError("Reply text cannot be empty.")
        self._reply = reply_text

    def set_id(self, new_id: int):
        """Allows repository to set the ID upon adding."""
        if not isinstance(new_id, int) or new_id <= 0:
             raise ValueError("New Enquiry ID must be a positive integer.")
        if self._enquiry_id != 0 and self._enquiry_id != new_id: # Allow setting if default 0
             print(f"Warning: Changing existing Enquiry ID from {self._enquiry_id} to {new_id}")
        self._enquiry_id = new_id

    def to_csv_dict(self) -> dict:
        return {
            'EnquiryID': self._enquiry_id,
            'ApplicantNRIC': self._applicant_nric,
            'ProjectName': self._project_name,
            'Text': self._text,
            'Reply': self._reply
        }

    @classmethod
    def from_csv_dict(cls, row_dict: dict) -> 'Enquiry':
        try:
            return cls(
                enquiry_id=int(row_dict['EnquiryID']),
                applicant_nric=row_dict['ApplicantNRIC'],
                project_name=row_dict['ProjectName'],
                text=row_dict['Text'],
                reply=row_dict.get('Reply', '')
            )
        except (KeyError, ValueError, TypeError) as e:
            raise DataLoadError(f"Error creating Enquiry from CSV row: {row_dict}. Error: {e}")

    def __eq__(self, other):
        if not isinstance(other, Enquiry): return NotImplemented
        return self._enquiry_id == other._enquiry_id

    def __hash__(self):
        return hash(self._enquiry_id)

    def get_display_summary(self) -> str:
         reply_status = "Replied" if self.is_replied() else "Unreplied"
         text_preview = (self._text[:47] + '...') if len(self._text) > 50 else self._text
         return f"ID: {self._enquiry_id:<4} | Project: {self._project_name:<15} | Status: {reply_status:<9} | Text: {text_preview}"

File: J2/Models/Enquiry.java
package Models;

import java.util.Date;
import java.util.Objects;
import java.util.concurrent.atomic.AtomicLong; // Use AtomicLong for thread safety

public class Enquiry {
    private final String enquiryId;
    private final String applicantNric;
    private final String projectName;
    private String enquiryText; // Mutable if not replied
    private String replyText; // Null until replied
    private String repliedByNric; // Null until replied
    private final Date enquiryDate;
    private Date replyDate; // Null until replied

    // Thread-safe static counter for generating unique IDs
    private static final AtomicLong nextId = new AtomicLong(1);

    // Constructor for new enquiries
    public Enquiry(String applicantNric, String projectName, String enquiryText, Date enquiryDate) {
        // Use Objects.requireNonNull for cleaner null checks
        Objects.requireNonNull(applicantNric, "Applicant NRIC cannot be null");
        Objects.requireNonNull(projectName, "Project Name cannot be null");
        Objects.requireNonNull(enquiryText, "Enquiry Text cannot be null");
        Objects.requireNonNull(enquiryDate, "Enquiry Date cannot be null");

        if (applicantNric.trim().isEmpty() || projectName.trim().isEmpty() || enquiryText.trim().isEmpty()) {
            throw new IllegalArgumentException("Enquiry fields (NRIC, Project, Text) cannot be empty");
        }

        // Generate unique ID atomically
        this.enquiryId = "ENQ" + nextId.getAndIncrement();
        this.applicantNric = applicantNric;
        this.projectName = projectName;
        this.enquiryText = enquiryText.trim(); // Trim input text
        this.enquiryDate = enquiryDate;
        // Initialize reply fields to null
        this.replyText = null;
        this.repliedByNric = null;
        this.replyDate = null;
    }

    // Constructor for loading from persistence
    public Enquiry(String enquiryId, String applicantNric, String projectName, String enquiryText, String replyText, String repliedByNric, Date enquiryDate, Date replyDate) {
        // Null checks for essential fields during load
        Objects.requireNonNull(enquiryId, "Enquiry ID cannot be null when loading");
        Objects.requireNonNull(applicantNric, "Applicant NRIC cannot be null when loading");
        Objects.requireNonNull(projectName, "Project Name cannot be null when loading");
        Objects.requireNonNull(enquiryText, "Enquiry Text cannot be null when loading");
        Objects.requireNonNull(enquiryDate, "Enquiry Date cannot be null when loading");

        this.enquiryId = enquiryId;
        this.applicantNric = applicantNric;
        this.projectName = projectName;
        this.enquiryText = enquiryText; // Assume already trimmed from source
        // Handle potentially empty/null reply fields from CSV
        this.replyText = (replyText == null || replyText.trim().isEmpty()) ? null : replyText.trim();
        this.repliedByNric = (repliedByNric == null || repliedByNric.trim().isEmpty()) ? null : repliedByNric.trim();
        this.enquiryDate = enquiryDate;
        this.replyDate = replyDate; // Can be null if not replied

        // Update the static counter based on loaded IDs
        updateNextId(enquiryId);
    }

    // Method to update the static ID counter based on loaded IDs
    // Should be called by the repository after loading all enquiries
    public static void updateNextId(String loadedEnquiryId) {
         if (loadedEnquiryId != null && loadedEnquiryId.startsWith("ENQ")) {
            try {
                // Extract number part after "ENQ"
                long idNum = Long.parseLong(loadedEnquiryId.substring(3));
                // Atomically set nextId if the loaded ID is greater or equal
                nextId.accumulateAndGet(idNum + 1, Math::max);
            } catch (NumberFormatException | StringIndexOutOfBoundsException e) {
                // Log warning if parsing fails, but don't stop loading
                System.err.println("Warning: Could not parse enquiry ID for nextId update: " + loadedEnquiryId);
            }
         }
    }

    // Method to call after all loading is done (e.g., in Repository or DataService)
     public static void finalizeNextIdInitialization() {
         System.out.println("Enquiry nextId initialized to: " + nextId.get());
     }

    // --- Getters ---
    public String getEnquiryId() { return enquiryId; }
    public String getApplicantNric() { return applicantNric; }
    public String getProjectName() { return projectName; }
    public String getEnquiryText() { return enquiryText; }
    public String getReplyText() { return replyText; }
    public String getRepliedByNric() { return repliedByNric; }
    public Date getEnquiryDate() { return enquiryDate; }
    public Date getReplyDate() { return replyDate; }

    // --- Setters ---

    /**
     * Sets the enquiry text. Can only be done if the enquiry has not been replied to.
     * @param enquiryText The new enquiry text (cannot be null or empty).
     * @return true if the text was set successfully, false otherwise.
     */
    public boolean setEnquiryText(String enquiryText) {
        if (isReplied()) {
             System.err.println("Cannot edit an enquiry that has already been replied to (ID: " + enquiryId + ").");
             return false;
        }
        if (enquiryText == null || enquiryText.trim().isEmpty()) {
             System.err.println("Enquiry text cannot be empty.");
             return false;
        }
        this.enquiryText = enquiryText.trim();
        return true;
    }

    /**
     * Sets the reply details for the enquiry.
     * @param replyText The reply text (cannot be null or empty).
     * @param repliedByNric The NRIC of the user who replied (cannot be null or empty).
     * @param replyDate The date of the reply (cannot be null).
     * @return true if the reply was set successfully, false otherwise.
     */
    public boolean setReply(String replyText, String repliedByNric, Date replyDate) {
        // Prevent setting reply if already replied? Or allow overwriting? Let's prevent.
        if (isReplied()) {
             System.err.println("Enquiry " + enquiryId + " has already been replied to. Cannot set reply again.");
             return false;
        }
        if (replyText == null || replyText.trim().isEmpty() ||
            repliedByNric == null || repliedByNric.trim().isEmpty() ||
            replyDate == null)
        {
            System.err.println("Invalid reply parameters provided for enquiry ID: " + enquiryId + ". All fields (text, replier NRIC, date) are required.");
            return false;
        }
        this.replyText = replyText.trim();
        this.repliedByNric = repliedByNric.trim();
        this.replyDate = replyDate;
        return true;
    }

    // --- State Check ---
    public boolean isReplied() {
        // Considered replied if replyText is not null and not empty
        return this.replyText != null && !this.replyText.isEmpty();
    }

    // --- Standard overrides ---
    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Enquiry enquiry = (Enquiry) o;
        return enquiryId.equals(enquiry.enquiryId); // ID is unique identifier
    }

    @Override
    public int hashCode() {
        return Objects.hash(enquiryId);
    }

    @Override
    public String toString() {
        return "Enquiry{" +
                "enquiryId='" + enquiryId + '\'' +
                ", applicantNric='" + applicantNric + '\'' +
                ", projectName='" + projectName + '\'' +
                ", enquiryDate=" + enquiryDate +
                ", replied=" + isReplied() +
                '}';
    }
}
